src/com/facebook/buck/jvm/kotlin/KotlinLibraryDescription.java

```diff
@@ -0,0 +1,129 @@
+import com.facebook.buck.rules.BuildRuleType;
+import com.facebook.buck.rules.BuildRules;
+import com.facebook.buck.rules.BuildTargetSourcePath;
+import com.facebook.buck.rules.Description;
+import com.facebook.buck.rules.SourcePath;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.rules.TargetGraph;
+import com.facebook.infer.annotation.SuppressFieldNotInitialized;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Iterables;
+
+import java.nio.file.Path;
+import java.util.regex.Pattern;
+
+
+public class KotlinLibraryDescription implements Description<KotlinLibraryDescription.Arg> {
+
+  public static final BuildRuleType TYPE = BuildRuleType.of("kotlin_library");
+
+  private final KotlinBuckConfig kotlinBuckConfig;
+
+  public KotlinLibraryDescription(KotlinBuckConfig kotlinBuckConfig) {
+    this.kotlinBuckConfig = kotlinBuckConfig;
+  }
+
+  @Override
+  public BuildRuleType getBuildRuleType() {
+    return TYPE;
+  }
+
+  @Override
+  public Arg createUnpopulatedConstructorArg() {
+    return new Arg();
+  }
+
+  @Override
+  public <A extends Arg> BuildRule createBuildRule(
+      TargetGraph targetGraph,
+      BuildRuleParams params,
+      BuildRuleResolver resolver,
+      A args) throws NoSuchBuildTargetException {
+    SourcePathResolver pathResolver = new SourcePathResolver(resolver);
+
+    BuildTarget abiJarTarget = params.getBuildTarget().withAppendedFlavors(CalculateAbi.FLAVOR);
+
+    ImmutableSortedSet<BuildRule> exportedDeps = resolver.getAllRules(args.exportedDeps.get());
+    DefaultJavaLibrary defaultJavaLibrary =
+        resolver.addToIndex(
+            new DefaultJavaLibrary(
+                params.appendExtraDeps(
+                    BuildRules.getExportedRules(
+                        Iterables.concat(
+                            params.getDeclaredDeps().get(),
+                            exportedDeps,
+                            resolver.getAllRules(args.providedDeps.get())))),
+                pathResolver,
+                args.srcs.get(),
+                validateResources(
+                    pathResolver,
+                    params.getProjectFilesystem(),
+                    args.resources.get()),
+                Optional.<Path>absent(),
+                Optional.<SourcePath>absent(),
+                ImmutableList.<String>of(),
+                exportedDeps,
+                resolver.getAllRules(args.providedDeps.get()),
+                new BuildTargetSourcePath(abiJarTarget),
+                /* trackClassUsage */ false,
+                /* additionalClasspathEntries */ ImmutableSet.<Path>of(),
+                new KotlincToJarStepFactory(
+                    kotlinBuckConfig.getKotlinCompiler().get(),
+                    args.extraKotlincArguments.or(ImmutableList.<String>of())),
```

Comment:
No need for the `or` here, the coercer will map an absent to an empty list: https://github.com/facebook/buck/blob/master/src/com/facebook/buck/rules/coercer/ListTypeCoercer.java#L39 .